Disclosure log to record ICS3 liveness problems (#83)
informalsystems/ibc-rs: docs/disclosure-log.md

# Disclosure Log for IBC Protocols

This document is a record of all the bugs or issues we uncovered while specifying & formally verifying the IBC protocols.

> ##### 1. ICS3 deadlock due to relayer algorithm

Reviewer: As I understand, this is a mistake in the example relayer algorithm described in ICS 18, not a mistake in ICS 3 itself - am I understanding correctly? Or is there code in ICS 3 which needs to be fixed as well?

Reviewer 2: Just needs fix in ICS-18

Author: Added a remark clarifying that only ICS018 requires modifications.

For completeness, I added a separate issue in cosmos/ics for tracking the required changes in ICS018: https://github.com/cosmos/ics/issues/436

The algorithm for relaying connection handshake datagrams of type `ConnOpenTry`does not handle the situation when both chains are in state `INIT`. The current relayer algorithm in [ICS018](https://github.com/cosmos/ics/tree/master/spec/ics-018-relayer-algorithms) specifies that the `ConnOpenTry` datagram should be relayed only if one of the chains is in state `INIT` and the other chain is uninitialized (see the snippet below) which can lead to a deadlock.

```
    if (localEnd.state === INIT && remoteEnd === null)
```

The correct code should include both the cases when a single chain is in state `INIT` and the case when both chains are in state `INIT`, as specified here: [Relayer.tla](https://github.com/informalsystems/ibc-rs/blob/master/docs/spec/relayer/Relayer.tla#L174)

##### 2. ICS3 deadlock due to `UpdateClient` semantics

Related issues: [#71](https://github.com/informalsystems/ibc-rs/issues/71) and [#61](https://github.com/informalsystems/ibc-rs/issues/61).

> Code: [#73](https://github.com/informalsystems/ibc-rs/pull/73).

Reviewer 2: Not sure what ICS3 deadlock means here, also above. In both cases it seems that we deal with a relayer that has some implementation bugs. Is this (#2) also about a relayer using ICS-18 algorithm? 

Here I believe we need a bit more detail on the consequences of a relayer not handling `UpdateClient` failure properly and describe when and how the deadlock can occur (e.g. presence of another relayer that somehow manages to do updates at precise times). Then we need to propose a change to ICS-18 like you do in #1.

Also let's try to close https://github.com/informalsystems/ibc-rs/pull/73 first before we merge this. Have you managed to show the deadlock here?

Reviewer: To me https://github.com/informalsystems/ibc-rs/issues/61 describes a race condition, which is quite a different thing than a "deadlock" ("deadlock" brings to mind badly written concurrent code using mutual exclusion locks that can reach a state from which it is impossible to progress, which is not what we're talking about here)

Reviewer 2: Yes, I agree "deadlock" is not the best word, we will rephrase. Also in 1. In both cases the handshake cannot progress. While in 1 it is clear why, here the scenarios for "cannot progress" have to be detailed.

Author: Hmm, by "deadlock" I meant, loosely, "loss of liveness." I agree this can be confusing.

To prevent confusion, I removed "deadlock" and replaced with "liveness problem."

Author: > While in 1 it is clear why, here the scenarios for "cannot progress" have to be detailed.

The "cannot progress" details are given in the referenced code (#73).
More specifically, see the "problematic execution" here: https://github.com/informalsystems/ibc-rs/blob/e33f32db95314461d18fbdb13574ddeabafedad4/verification/spec/connection-handshake/updateclient-deadlock/README.md#specification-details

I'm not sure it's worth detailing here (in addition to the execution in #73). @ancazamfir but if you think it's relevant for completeness sake, then I can describe the issue.

The current relayer algorithm in [ICS018](https://github.com/cosmos/ics/tree/master/spec/ics-018-relayer-algorithms) specifies that the `ConnOpenTry` datagram should be relayed only if one of the chains is in state `INIT` and the other chain is uninitialized (see the snippet below); this is not enough for guaranteeing liveness of the connection handshake protocol (ICS04).

```
    if (localEnd.state === INIT && remoteEnd === null)
```

The correct code should include both the cases when a single chain is in state `INIT`, as well as the case when both chains are in state `INIT`, as specified here: [Relayer.tla](https://github.com/informalsystems/ibc-rs/blob/master/docs/spec/relayer/Relayer.tla#L174)
This fix only concerns the relayer algorithm ICS018.

###### Channel handshake (ICS4) liveness problem

The same issue (and fix) seems to exist for the channel handshake datagrams.

##### 2. ICS3 liveness problem due to `UpdateClient` semantics

Related issues: [#71](https://github.com/informalsystems/ibc-rs/issues/71) and [#61](https://github.com/informalsystems/ibc-rs/issues/61).

Code for obtaining an execution trace in TLA+: [#73](https://github.com/informalsystems/ibc-rs/pull/73).

> Briefly, this trace comprises the following sequence of actions, assuming two chains `A` and `B` and a relayer (as a simplification, the chains in this model are in charge of creating messages, a responsibility that the relayer typically implements):

Reviewer 3: This sentence is confusing as it says that modules sent directly messages and then there is a relayer also. In case we want to stay in the model of #73 we can say we have a model in which modules send messages to each other but network can also inject faulty messages (simulating concurrent correct and faulty relayers). Then scenario is explained where network inject UpdateClient message with h+1, which leads to failure of M to be delivered.

> 0. Chain `A` creates a message `M` that has a proof for height `h`, chain `A` also advances its height to `h+1`.

Reviewer 2: ```suggestion
0. Chain `A` creates a message `M` that includes a proof at height `h`, chain `A` also advances its height to `h+1`.
```

> 1. The relayer updates the client on chain `B` with the latest height of chain A, namely `h+1` (call to `UpdateClient`).

Reviewer 2: ```suggestion
1. Another relayer updates the client on chain `B` with the latest height of chain A, namely `h+1` (call to `UpdateClient`).
```

Author: As described below, this trace does not comprise multiple relayers. The loss of liveness happens due to solely one relayer and the way `updateClient` is used.

Reviewer 3: It is more complicated explaining this scenario with only single relayer as it would not update client unless there is a corresponding message to send. The exception would be the case where the highest header is about to expire. But this is not the scenario explained here. Maybe the confusion come from the fact that in the used model we don't have explicit relayer, and relayer is being used to describe the problem.

> 2. The relayer sends message `M` that chain `A` created (at step 0) to chain B, and also attempt to update the client on chain `B` with height `h` (the height in the proof of this message), but is unable to do so because `h` is smaller than `h +1` (the latest height of client on chain B).

Reviewer 2: ```suggestion
2. The relayer sends two messages to chain `B`, possibly in a single transaction:
- an `updateClient` message for height `h` and,
- message `M` that was vreated created at step 0
But when processed on `B` the `updateClient(h)` fails as the client's state is at a higher height (`h+1`) than `h`.
```

Author: >  The relayer sends two messages to chain `B`, possibly in a single transaction:

This sentence is not really accurate w.r.t to the trace in #73. (There is no `updateClient` __message__, as updates do not rely on messages.)

> 3. Chain `B` will be unable to verify message `M` since it lacks the adequate height information `h`, so chain `B` will drop this message, hence ICS3 can no longer progress.

Reviewer 2: ```suggestion
3. Chain `B`is unable to verify message `M` since it lacks the consensus state at height `h`. Chain `B` drops this message, hence ICS3 can no longer progress.
```

Author: > It is possible that chain B has a consensus state at height h from a previous update by the "other" relayer. So even if the updateClient(h) failed above, M could still be accepted.

That's true.. but this is not really the scenario in this execution. The trace here has a single relayer. I'm not sure if it helps to add more information about other potential problematic scenarios involving multiple relayers.

We classify this under ICS3, however, since this is the context where we discovered the problem.
The TLA+ spec we have for depicting this liveness problem is for the ICS3 protocol.

##### Problem statement

Related issues: [#71](https://github.com/informalsystems/ibc-rs/issues/71) and [#61](https://github.com/informalsystems/ibc-rs/issues/61).
The problem is more thorougly described in #61, but for completeness sake we restated it here in a compact form.

The liveness property that a correct relayer should provide is eventual delivery.
Assuming some source chain `A`, destination chain `B`, and an IBC item `X` (e.g., connection, channel, or packet) on chain `A`, we can define this property as follows:

> For any IBC item `X` on chain `A` destined for chain `B`, eventually, a correct relayer will submit item `X` to chain `B`.

This is difficult to guarantee in practice, however.
Intuitively, the difficulty arises because of a combination of two factors:

1. __Proof requirement:__ For chain `B` to accept item `X`, it must verify the authenticity of this item; this is done via the light client that chain `B` maintains.

> This light client requires, given item `X` at height `h` on `A`, a proof constructed at height `h-1`.
> Put simply, the light client needs: (1) the item `X` at height `h` (as part of what is called "consensus state"), plus (2) a proof constructed at height `h-1` for this item.

Reviewer 2: ```suggestion
Given an item `X` and a commitment proof for `X` constructed at height `h-1`, the light client requires the consensus state at height `h` that includes that commitment root required for verification.
```

> 2. __Concurrency:__ Different relayers may update the same light client. Suppose a relayer `r1` wants to submit `X` at consensus state `h` plus the associated proof at `h-1`.

Reviewer 2: ```suggestion
2. __Concurrency:__ Different relayers may update the same light client. Suppose a relayer `r1` wants to submit a consensus state at height `h` and a message that includes `X` plus the associated proof at `h-1`.
```

In the meantime, however, another relayer `r2` may update this same light client to height `h'`.

> It is important to note that `h'` is bigger than `h`.

Reviewer 2: ```suggestion
Assume `h'` is bigger than `h`.
```

Now `r1` will be unable to update the light client with consensus state `h`, because the light client disallows updates with smaller heights than the current height `h'`; consequently, the relayer will be unable to submit `X`.

> To ensure eventual delivery, relayer `r1` would need to retry submitting item `X`, that is: resubmit the item at a larger height (e.g., at `h'`) with the associated proof (at `h'-1`).

Reviewer 2: ```suggestion
To ensure eventual delivery, relayer `r1` would need to retry submitting item `X`, that is: resubmit the consensus state at a larger height (e.g., at `h'`) followed by the message that includes the proof for `X` (at `h'-1`).
```

The liveness property that a correct relayer should provide is eventual delivery.
Assuming some source chain `A`, destination chain `B`, and an IBC item `X` (e.g., connection, channel, or packet) on chain `A`, we can define this property as follows:

> For any IBC item `X` on chain `A` destined for chain `B`, eventually, a correct relayer will submit item `X` to chain `B`.

This is difficult to guarantee in practice, however.
Intuitively, the difficulty arises because of a combination of two factors:

1. __Proof requirement:__ For chain `B` to accept item `X`, it must verify the authenticity of this item; this is done via the light client that chain `B` maintains.
Given an item `X` and a commitment proof for `X` constructed at height `h-1`, the light client requires the consensus state at height `h` that includes that commitment root required for verification.

2. __Concurrency:__ Different relayers may update the same light client.
Suppose a relayer `r1` wants to submit a consensus state at height `h`.
In the meantime, however, another relayer `r2` may update this same light client to height `h'`.
Assume `h'` is bigger than `h`.

> Now `r1` will be unable to update the light client with consensus state `h`, because the light client disallows updates with smaller heights than the current height `h'`.

Reviewer 2: This is true for TM light client. Might be different for others.
```suggestion
If the light client disallows updates with heights smaller than the current height `h'` then `r1`'s update fails .
```

> Consequently, the relayer will be unable to submit consensus state at height `h`.

Reviewer 2: ```suggestion
Consequently, the relayer will be unable to submit message for `X`.
```

Author: `X` is not mentioned in (2/Concurrency), so it would be confusing to talk about it here.
